### DISCUSSION TITLE
How to use automatic differentiation correctly

### DISCUSSION THREAD

**Discussion by nuomi68** [idaholab/moose#24583] [Q&A Tools] [answered]
Hello, everyone！
I want to use automatic differentiation in gap heat transfer,I'm trying to create some values with ADReal, but I don't know how to get their derivatives.
I tried modifying the code for ` GapConductance`
```
void
GasGapConductance::computeQpConductance()
{
  if (_has_info)
  {
    const ADReal cache_conductance= h_conduction() + h_radiation()+h_contact();
    // _gap_conductance[_qp] = h_conduction() + h_radiation();
     _gap_conductance[_qp] = cache_conductance.value();
    // _gap_conductance_dT[_qp] = dh_conduction() + dh_radiation();
    _gap_conductance_dT[_qp] =cache_conductance.derivatives();
  }
  else
  {
    _gap_conductance[_qp] = 0;
    _gap_conductance_dT[_qp] = 0;
  }
}
```
I got an error like this
`error: cannot convert 'const MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<500> >' to 'MooseADWrapper<double, false>' {aka 'double'} in assignment
                  _gap_conductance_dT[_qp] =cache_conductance.derivatives();` 

I hope to get your help!

**Comment by nuomi68** [comment]
For example, for h_radiation(), I added AD keyword, compared with the original MOOSE
```
ADReal
GasGapConductance::h_radiation()
{
  if (_emissivity == 0.0)
    return 0.0;

  ADReal surface_integration_factor = 1.0;

  if (_gap_geometry_type == GasGapConductance::CYLINDER)
    surface_integration_factor = 0.5 * (_r1 + _r2) / _radius;
  else if (_gap_geometry_type == GasGapConductance::SPHERE)
    surface_integration_factor = 0.25 * (_r1 + _r2) * (_r1 + _r2) / (_radius * _radius);

  const ADReal temp_func =
      (_temp[_qp] * _temp[_qp] + _gap_temp * _gap_temp) * (_temp[_qp] + _gap_temp);

  return _stefan_boltzmann * temp_func / _emissivity * surface_integration_factor;
}
```

**Comment by GiudGiud** [comment, marked as answer]
```
error: cannot convert 'const MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<500> >' to 'MooseADWrapper<double, false>' {aka 'double'} in assignment _gap_conductance_dT[_qp] =cache_conductance.derivatives();
```

means you are trying to implicitly convert from an ADReal to a Real.
The material property is probably declared as a Real

**Reply by nuomi68** [reply]
Yeah, I'd like to compute the derivative of cache_conductance for temperature. But the derivative expression displayed is more complex, so I want to use automatic differentiation. Whether there are reference cases?

**Reply by GiudGiud** [reply]
What does the header look like right now?

**Reply by nuomi68** [reply]
Thank you for your attention, I have solved my problem now.

**Comment by lindsayad** [comment]
AD residual objects should never need to access the `.derivatives()` API. This data member `_gap_conductance_dT` should go away for example

**Reply by lindsayad** [reply]
`_gap_conductance` should be an `ADReal` and should contain both the function value and its derivatives. Yea `.derivatives()` and `.value()` should not be called

**Reply by nuomi68** [reply]
Thank you for your answer, I think I have achieved what I wanted to do, maybe not the recommended way to achieve. I used ChainedReal,

```
const GenericChainedReal<true> cache_temp =GenericChainedReal<true>(_temp[_qp], 1);
 const GenericChainedReal<true> cache_conductance=  h_radiation(cache_temp);
_gap_conductance[_qp] =  cache_conductance.value().value();
_gap_conductance_dT[_qp] =cache_conductance.derivatives().value();
```
```
GenericChainedReal<true>
GasGapConductance::h_radiation(const GenericChainedReal<true> & temp)
{
  if (_emissivity == 0.0)
    return GenericChainedReal<true>(0, 0);
  ADReal surface_integration_factor = 1.0;
  const GenericChainedReal<true> temp_func =
      (temp * temp + _gap_temp * _gap_temp) * (temp + _gap_temp);
  return _stefan_boltzmann * temp_func / _emissivity * surface_integration_factor;
}
```
I inserted a little AD calculation in the non-AD version because the derivative expression of gap_conductance for temperature is too complex.I used a simple example to do the experiment, and it was numerically correct，`GapFluxModelPressureDependentConduction` module seems to also have this kind of insert AD
